Email Folders Don't Select

<!--- Provide a general summary of the issue in the **Title** above -->
<!--- Before you open an issue, please check if a similar issue already exists or has been closed before. --->
<!--- If you have discovered a security risk please report it by emailing security@suitecrm.com. This will be delivered to the product team who handle security issues. Please don't disclose security bugs publicly until they have been handled by the security team. --->
<!--- Please be aware that as of the 31st January 2022 we no longer support 7.10.x.
New issues referring to 7.10.x will only be valid if applicable to 7.12.x and above.
If your issue is still applicable in 7.12.x, please create the issue following the template below -->
#### Issue
<!--- Provide a more detailed introduction to the issue itself, and why you consider it to be a bug -->
<!--- Ensure that all code ``` is surrounded ``` by triple back quotes. This can also be done over multiple lines -->
In the emails view, the IMAP account folders don't select. They always default back to the INBOX.
#### Expected Behavior
<!--- Tell us what should happen -->
The IMAP folder should be selected and displayed.

#### Actual Behavior
<!--- Tell us what happens instead -->
<!--- Also please check relevant logs (suitecrm.log, php error.log etc.) -->
When selecting an email folder, the program does not select it and display its contents. Instead, it defaults back to INBOX. The resulting log entries are as follows:

```Tue Apr 18 19:07:07 2023 [167845][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] User::getPreference() should not be used statically.
Tue Apr 18 19:07:07 2023 [167845][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] User::getPreference() should not be used statically.
Tue Apr 18 19:07:07 2023 [167845][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] User::getPreference() should not be used statically.
Tue Apr 18 19:07:07 2023 [167845][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] Array
(
)

Tue Apr 18 19:07:08 2023 [167885][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] Array
(
)

Tue Apr 18 19:07:10 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] Array
(
)

Tue Apr 18 19:07:10 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][WARN] CSS File Day/yui.css not found
Tue Apr 18 19:07:10 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][ERROR] Unable to find relationship emails_email_templates
Tue Apr 18 19:07:10 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][WARN] emails_email_templates_idb for emails_email_templates failed to load

Tue Apr 18 19:07:10 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][WARN] Email ID is Empty
Tue Apr 18 19:07:10 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] Using row number in fetchByAssoc is not portable and no longer supported. Please fix your code.
Tue Apr 18 19:07:10 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][WARN] Trying to get property of non-object: list view data "order by" is not defined
Tue Apr 18 19:07:10 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][WARN] Trying to get property of non-object: list view data "sort order" is not defined
Tue Apr 18 19:07:10 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] Using row number in fetchByAssoc is not portable and no longer supported. Please fix your code.
Tue Apr 18 19:07:11 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] User::getPreference() should not be used statically.
Tue Apr 18 19:07:11 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] User::getPreference() should not be used statically.
Tue Apr 18 19:07:11 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] User::getPreference() should not be used statically.
Tue Apr 18 19:07:11 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][WARN] unknown or undefined folder type (we will use 'imap' instead): inbound
Tue Apr 18 19:07:11 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][WARN] Request ssl value not found.
Tue Apr 18 19:07:11 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][WARN] ImapHandler trying to use a non valid resource stream.
Tue Apr 18 19:07:11 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][WARN] ImapHandler trying to use a non valid resource stream.
Tue Apr 18 19:07:11 2023 [168051][7356b911-2136-7ab5-05b1-642dc5a54da7][WARN] $value must be a string
Tue Apr 18 19:07:12 2023 [168120][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] User::getPreference() should not be used statically.
Tue Apr 18 19:07:12 2023 [168120][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] User::getPreference() should not be used statically.
Tue Apr 18 19:07:12 2023 [168120][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] User::getPreference() should not be used statically.
Tue Apr 18 19:07:12 2023 [168120][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] Array
(
)

Tue Apr 18 19:07:13 2023 [168206][7356b911-2136-7ab5-05b1-642dc5a54da7][DEPRECATED] Array
(
)

Tue Apr 18 19:07:13 2023 [168206][7356b911-2136-7ab5-05b1-642dc5a54da7][WARN] CSS File Day/yui.css not found
```

#### Possible Fix
<!--- Not obligatory, but suggest a fix or reason for the bug -->

#### Steps to Reproduce
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug include code to reproduce, if relevant -->

#### Context
<!--- How has this bug affected you? What were you trying to accomplish? -->
<!--- If you feel this should be a low/medium/high priority then please state so -->
It is difficult to tell if emails have in fact been sent, and obviously one cannot navigate the email account correctly.

#### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* SuiteCRM Version used: 7.13.2
* Browser name and version (e.g. Chrome Version 51.0.2704.63 (64-bit)): Chrome Version 112.0.5615.121 (Official Build) (x86_64)
* Environment name and version (e.g. MySQL, PHP 7): Blue Host shared server, cPanel 102.0, Apache 2.4.57, PHP 8.1.17, MySQL 5.7.23
* Operating System and version (e.g Ubuntu 16.04): CentOS Linux

--- a/include/Imap/Imap2Handler.php
+++ b/include/Imap/Imap2Handler.php
@@ -1125,7 +1125,7 @@ class Imap2Handler implements ImapHandlerInterface
 
         if ($messageSet === null && $client->getCapability('SORT') !== false) {
 
-            $result = $client->sort($this->getStream()->getMailboxName(), 'REVERSE ' . $field, "$start:$end",
+            $result = $client->sort($this->getStream()->getMailboxName(), 'REVERSE ' . $field, "UNDELETED $start:$end",
                 $returnUid, 'UTF-8');
 
             if (empty($result) || empty($result->count())) {

--- a/include/MassUpdate.php
+++ b/include/MassUpdate.php
@@ -112,7 +112,7 @@ class MassUpdate
         unset($_REQUEST['current_query_by_page']);
         unset($_REQUEST[session_name()]);
         unset($_REQUEST['PHPSESSID']);
-        $query = json_encode($_REQUEST);
+        $query = htmlspecialchars(json_encode($_REQUEST), ENT_QUOTES, 'UTF-8');
 
         if (!isset($_REQUEST['module'])) {
             LoggerManager::getLogger()->warn('Undefined index: module');

--- a/include/SugarFolders/SugarFolders.php
+++ b/include/SugarFolders/SugarFolders.php
@@ -884,7 +884,7 @@ class SugarFolder
                 }
 
                 if ($a['has_child'] == 1) {
-                    $qGetChildren = $this->core . $this->coreWhere . "AND parent_folder = " . $this->db->quoted($a['id']);
+                    $qGetChildren = $this->core . $this->coreWhere . "AND parent_folder = " . $this->db->quoted($a['id']) . $this->coreOrderBy;
                     $rGetChildren = $this->db->query($qGetChildren);
 
                     while ($aGetChildren = $this->db->fetchByAssoc($rGetChildren)) {
@@ -968,7 +968,7 @@ class SugarFolder
 
         if ($a['has_child'] == 1) {
             $this->_depth++;
-            $qGetChildren = $this->core . $this->coreWhere . " AND parent_folder = " . $this->db->quoted($a['id']);
+            $qGetChildren = $this->core . $this->coreWhere . "AND parent_folder = " . $this->db->quoted($a['id']) . $this->coreOrderBy;
             $rGetChildren = $this->db->query($qGetChildren);
 
             while ($aGetChildren = $this->db->fetchByAssoc($rGetChildren)) {
@@ -1060,7 +1060,7 @@ class SugarFolder
             $folderNode->set_property('children', array());
 
             if (in_array($a['id'], $subscriptions) && $a['has_child'] == 1) {
-                $qGetChildren = $this->core . $this->coreWhere . "AND parent_folder = " . $this->db->quoted($a['id']);
+                $qGetChildren = $this->core . $this->coreWhere . "AND parent_folder = " . $this->db->quoted($a['id']) . $this->coreOrderBy;
                 $rGetChildren = $this->db->query($qGetChildren);
 
                 while ($aGetChildren = $this->db->fetchByAssoc($rGetChildren)) {

--- a/include/SugarFolders/SugarFolders.php
+++ b/include/SugarFolders/SugarFolders.php
@@ -1541,10 +1541,41 @@ class SugarFolder
             if (isFalse($isSelected)) {
                 continue;
             }
-            $id = $folder['id'] ?? '';
 
-            if ($id === $folderId) {
+            if (($folder['id'] ?? '') === $folderId) {
                 return true;
+            }
+
+            if ($folder['has_child'] == 1) {
+                $children = $folder['children'] ?? [];
+                if (!empty($children) && $this->folderExistsInSubtree($children, $folderId)) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Recursively search for a folder ID within a subtree without applying
+     * the 'selected' filter — child folders don't carry that property.
+     *
+     * @param array $folders
+     * @param string $folderId
+     * @return bool
+     */
+    private function folderExistsInSubtree(array $folders, string $folderId): bool
+    {
+        foreach ($folders as $folder) {
+            if (($folder['id'] ?? '') === $folderId) {
+                return true;
+            }
+            if ($folder['has_child'] == 1) {
+                $children = $folder['children'] ?? [];
+                if (!empty($children) && $this->folderExistsInSubtree($children, $folderId)) {
+                    return true;
+                }
             }
         }
 

--- a/modules/Emails/EmailsController.php
+++ b/modules/Emails/EmailsController.php
@@ -702,6 +702,26 @@ class EmailsController extends SugarController
     }
 
     /**
+     * @param InboundEmail $inboundEmail
+     * @param array $request
+     */
+    private function setMailboxFromRequest(InboundEmail $inboundEmail, array $request): void
+    {
+        if (!empty($request['folders_id'])) {
+            include_once 'modules/Emails/Folder.php';
+            try {
+                $folder = new Folder();
+                $folder->retrieve($request['folders_id']);
+                $inboundEmail->mailbox = $folder->getMailbox();
+            } catch (\SuiteException $e) {
+                LoggerManager::getLogger()->warn('setMailboxFromRequest: invalid folders_id — ' . $e->getMessage());
+            }
+        } elseif (!empty($request['folder']) && !in_array($request['folder'], ['sent', 'inbound'], true)) {
+            $inboundEmail->mailbox = html_entity_decode($request['folder'], ENT_QUOTES | ENT_HTML5, 'UTF-8');
+        }
+    }
+
+    /**
      * @throws SugarControllerException
      */
     public function action_DeleteFromImap()
@@ -720,6 +740,8 @@ class EmailsController extends SugarController
         }
 
         $inboundEmail = BeanFactory::getBean('InboundEmail', $db->quote($emailID));
+
+        $this->setMailboxFromRequest($inboundEmail, $_REQUEST);
 
         if (is_array($uid)) {
             $uid = implode(',', $uid);
@@ -759,7 +781,7 @@ class EmailsController extends SugarController
         // and select the folder
 
         $ie = $this->getInboundEmail($request['inbound_email_record']);
-        $ie->mailbox = $request['folder'];
+        $this->setMailboxFromRequest($ie, $request);
         $ie->connectMailserver();
 
         // get requested UIDs and flag type

--- a/modules/Emails/Folder.php
+++ b/modules/Emails/Folder.php
@@ -97,8 +97,17 @@ class Folder
     {
         $isValidator = new SuiteValidator();
         if ($isValidator->isValidId($folderId)) {
-            $result = $this->db->query("SELECT * FROM folders WHERE id='" . $folderId . "'");
+            $result = $this->db->query(
+                "SELECT f.*, ie.primary_folder, ie.mailbox AS ie_mailbox" .
+                " FROM folders f" .
+                " LEFT JOIN inbound_email ie ON ie.id = f.id AND ie.deleted = 0" .
+                " WHERE f.id='" . $folderId . "'"
+            );
             $row = $this->db->fetchByAssoc($result);
+
+            if (empty($row)) {
+                return null;
+            }
 
             // get the root of the tree
             // is the id of the root node is the same as the inbound email id
@@ -106,7 +115,13 @@ class Folder
                 // root node (inbound)
                 $this->id = $row['id'];
                 $this->type = $row['folder_type'];
-                $this->mailbox = 'INBOX'; // Top level IMAP folder
+
+                if (!empty($row['primary_folder'])) {
+                    $this->mailbox = $row['primary_folder'];
+                } else {
+                    $mailboxes = array_filter(array_map('trim', explode(',', $row['ie_mailbox'] ?? '')));
+                    $this->mailbox = !empty($mailboxes) ? reset($mailboxes) : 'INBOX';
+                }
             } else {
                 // child node
                 $this->id = $row['parent_folder'];

--- a/modules/Emails/Folder.php
+++ b/modules/Emails/Folder.php
@@ -117,16 +117,16 @@ class Folder
                 $this->type = $row['folder_type'];
 
                 if (!empty($row['primary_folder'])) {
-                    $this->mailbox = $row['primary_folder'];
+                    $this->mailbox = html_entity_decode($row['primary_folder'], ENT_QUOTES | ENT_HTML5, 'UTF-8');
                 } else {
                     $mailboxes = array_filter(array_map('trim', explode(',', $row['ie_mailbox'] ?? '')));
-                    $this->mailbox = !empty($mailboxes) ? reset($mailboxes) : 'INBOX';
+                    $this->mailbox = !empty($mailboxes) ? html_entity_decode(reset($mailboxes), ENT_QUOTES | ENT_HTML5, 'UTF-8') : 'INBOX';
                 }
             } else {
                 // child node
                 $this->id = $row['parent_folder'];
                 $this->type = $row['folder_type'];
-                $this->mailbox = $row['name'];
+                $this->mailbox = html_entity_decode($row['name'], ENT_QUOTES | ENT_HTML5, 'UTF-8');
             }
         } else {
             throw new SuiteException("Invalid or empty Email Folder ID");

--- a/modules/Emails/include/ListView/ListViewDataEmails.php
+++ b/modules/Emails/include/ListView/ListViewDataEmails.php
@@ -776,6 +776,13 @@ class ListViewDataEmails extends ListViewData
                         $offset
                     );
 
+                    // Inject folder info into $_REQUEST so MassUpdate::displayMassUpdate()
+                    // encodes it into the current_query_by_page hidden field, which
+                    // ListViewHeader.js reads to display the current folder name.
+                    $_REQUEST['folder'] = $folderObj->getMailbox() ?: $folder;
+                    $_REQUEST['folder_type'] = $folderObj->getType();
+                    $_REQUEST['inbound_email_record'] = $inboundEmail->id;
+
                     break;
 
                 case 'imap':

--- a/modules/Emails/include/ListView/ListViewHeader.js
+++ b/modules/Emails/include/ListView/ListViewHeader.js
@@ -139,11 +139,19 @@ $(document).ready(function () {
 
     for (let i = 0; i < (responses.length); i++) {
       if (responses[i].id === query.folders_id) {
-        jQueryBtnEmailsCurrentFolder.text(responses[(i)].text);
+        let folderText = $('<span>').html(responses[(i)].text).text();
+        jQueryBtnEmailsCurrentFolder.text(folderText);
       } else if (responses[i].id === query.inbound_email_record) {
         let regExp = /\(([^)]+)\)/;
-        let match = regExp.exec(responses[(i)].text);
-        jQueryBtnEmailsCurrentFolder.text(query.folder + ' ' + match[0]);
+        let decodedText = $('<span>').html(responses[(i)].text).text();
+        let match = regExp.exec(decodedText);
+        let folderEscaped = $('<span>').text(query.folder)[0].innerHTML;
+        let matchEscaped = match ? $('<span>').text(match[0])[0].innerHTML : '';
+        if (matchEscaped) {
+          jQueryBtnEmailsCurrentFolder.html(folderEscaped + ' ' + matchEscaped);
+        } else {
+          jQueryBtnEmailsCurrentFolder.html(folderEscaped);
+        }
       }
     }
   });

--- a/modules/Emails/metadata/listviewdefs.php
+++ b/modules/Emails/metadata/listviewdefs.php
@@ -49,7 +49,7 @@ $viewdefs['Emails']['ListView'] = array(
                         'customCode' => '<a class="btn" data-action="emails-show-compose-modal" title="{$MOD.LBL_COMPOSEEMAIL}"><span class="glyphicon glyphicon-envelope"></span></a>'
                     ),
                     array(
-                        'customCode' => '<a class="btn" data-action="emails-configure" title="{$MOD.LBL_EMAILSETTINGS}"><span class="glyphicon glyphicon-cog"></span></a>'
+                        'customCode' => '<a class="btn" href="index.php?module=InboundEmail&action=index" title="{$MOD.LBL_EMAILSETTINGS}"><span class="glyphicon glyphicon-cog"></span></a>'
                     ),
                     array(
                         'customCode' => '<a class="btn" data-action="emails-check-new-email" title="{$MOD.LBL_BUTTON_CHECK_TITLE}"><span class="glyphicon glyphicon-refresh"></span></a>'

--- a/modules/InboundEmail/InboundEmail.php
+++ b/modules/InboundEmail/InboundEmail.php
@@ -3129,14 +3129,14 @@ class InboundEmail extends SugarBean
      * @param int $id
      * @return int
      */
-    private function createFolder($name, $type, $focusUser, $id = 0)
+    public function createFolder($name, $type, $focusUser, $id = 0)
     {
         $folder = new SugarFolder();
         $folder->name = $name;
         $folder->folder_type = $type;
         $folder->has_child = $id ? 1 : 0;
         $folder->is_dynamic = 1;
-        $folder->dynamic_query = $this->generateDynamicFolderQuery("sent", $focusUser->id);
+        $folder->dynamic_query = $this->generateDynamicFolderQuery($type, $focusUser->id);
         $folder->created_by = $focusUser->id;
         $folder->modified_by = $focusUser->id;
 
@@ -3153,12 +3153,28 @@ class InboundEmail extends SugarBean
     }
 
     /**
-     * @param $folderName
+     * @param string $folderName
+     * @param string|null $trashFolder Explicit trash folder name; falls back to $_REQUEST['trashFolder']
+     * @param string|null $sentFolder  Explicit sent folder name;  falls back to $_REQUEST['sentFolder']
      * @return bool
      */
-    private function folderIsRequestTrashOrSent($folderName)
+    public function folderIsRequestTrashOrSent($folderName, $trashFolder = null, $sentFolder = null)
     {
-        return $folderName == $_REQUEST['trashFolder'] || $folderName == $_REQUEST['sentFolder'];
+        $trashFolder = $trashFolder ?? ($_REQUEST['trashFolder'] ?? '');
+        $sentFolder  = $sentFolder  ?? ($_REQUEST['sentFolder']  ?? '');
+
+        // Check case-insensitively against the explicitly configured special folders.
+        if (!empty($trashFolder) && strcasecmp($folderName, $trashFolder) === 0) {
+            return true;
+        }
+        if (!empty($sentFolder) && strcasecmp($folderName, $sentFolder) === 0) {
+            return true;
+        }
+
+        // Also guard against common IMAP system folder names that should never become
+        // regular inbound subfolders (matches the pattern used in the legacy migration code).
+        $systemFolders = ['sent', 'trash'];
+        return in_array(strtolower($folderName), $systemFolders, true);
     }
 
     /**

--- a/modules/InboundEmail/InboundEmail.php
+++ b/modules/InboundEmail/InboundEmail.php
@@ -3163,18 +3163,14 @@ class InboundEmail extends SugarBean
         $trashFolder = $trashFolder ?? ($_REQUEST['trashFolder'] ?? '');
         $sentFolder  = $sentFolder  ?? ($_REQUEST['sentFolder']  ?? '');
 
-        // Check case-insensitively against the explicitly configured special folders.
-        if (!empty($trashFolder) && strcasecmp($folderName, $trashFolder) === 0) {
+        if (!empty($trashFolder) && $folderName === $trashFolder) {
             return true;
         }
-        if (!empty($sentFolder) && strcasecmp($folderName, $sentFolder) === 0) {
+        if (!empty($sentFolder) && $folderName === $sentFolder) {
             return true;
         }
 
-        // Also guard against common IMAP system folder names that should never become
-        // regular inbound subfolders (matches the pattern used in the legacy migration code).
-        $systemFolders = ['sent', 'trash'];
-        return in_array(strtolower($folderName), $systemFolders, true);
+        return false;
     }
 
     /**
@@ -7041,7 +7037,9 @@ class InboundEmail extends SugarBean
         $this->filter_domain = $storedOptions['filter_domain'] ?? '';
         $this->trashFolder =  $storedOptions['trashFolder'] ?? '';
         $this->sentFolder = $storedOptions['sentFolder'] ?? '';
-        $this->mailbox = $storedOptions['mailbox'] ?? '';
+        if (!empty($storedOptions['mailbox'])) {
+            $this->mailbox = $storedOptions['mailbox'];
+        }
 
         $this->leave_messages_on_mail_server = isTrue($storedOptions['leaveMessagesOnMailServer'] ?? false);
         $this->move_messages_to_trash_after_import = !isTrue($storedOptions['leaveMessagesOnMailServer'] ?? true);

--- a/modules/InboundEmail/InboundEmail.php
+++ b/modules/InboundEmail/InboundEmail.php
@@ -6488,7 +6488,7 @@ class InboundEmail extends SugarBean
 
         if ($requestFolder === 'inbound') {
             if (!empty($_REQUEST['folder_name'])) {
-                $this->mailbox = $_REQUEST['folder_name'];
+                $this->mailbox = html_entity_decode($_REQUEST['folder_name'], ENT_QUOTES | ENT_HTML5, 'UTF-8');
             } elseif ($this->mailboxarray && (is_countable($this->mailboxarray) ? count($this->mailboxarray) : 0)) {
                 $this->mailbox = $this->mailboxarray[0];
             } else {
@@ -6496,7 +6496,11 @@ class InboundEmail extends SugarBean
             }
         }
 
-        $connectString = $this->getConnectString($service, $this->mailbox);
+        $encodedMailbox = $this->mailbox;
+        if (function_exists('mb_convert_encoding') && in_array('UTF7-IMAP', mb_list_encodings())) {
+            $encodedMailbox = mb_convert_encoding($this->mailbox, 'UTF7-IMAP', 'UTF-8') ?: $this->mailbox;
+        }
+        $connectString = $this->getConnectString($service, $encodedMailbox);
 
         /*
          * Try to recycle the current connection to reduce response times

--- a/modules/InboundEmail/Save.php
+++ b/modules/InboundEmail/Save.php
@@ -357,16 +357,6 @@ if (empty($foldersFoundRow)) {
             'created_by' => $focusUser->id,
             'modified_by' => $focusUser->id,
         ),
-        // Sent Emails
-        "sent" => array(
-            'name' => $mod_strings['LNK_SENT_EMAIL_LIST'] . ' ('.$stored_options['sentFolder'].')',
-            'folder_type' => "sent",
-            'has_child' => 0,
-            'dynamic_query' => '',
-            'is_dynamic' => 1,
-            'created_by' => $focusUser->id,
-            'modified_by' => $focusUser->id,
-        ),
         // Archived Emails
         "archived" => array(
             'name' => $mod_strings['LBL_LIST_TITLE_MY_ARCHIVES'],
@@ -408,6 +398,30 @@ if (empty($foldersFoundRow)) {
             $folder->save();
         }
     }
+
+    if (!empty($stored_options['sentFolder'])) {
+        $sentFolder = new SugarFolder();
+        $sentFolder->name = $mod_strings['LNK_SENT_EMAIL_LIST'] . ' ('.$stored_options['sentFolder'].')';
+        $sentFolder->folder_type = 'sent';
+        $sentFolder->has_child = 0;
+        $sentFolder->dynamic_query = '';
+        $sentFolder->is_dynamic = 1;
+        $sentFolder->created_by = $focusUser->id;
+        $sentFolder->modified_by = $focusUser->id;
+        $sentFolder->parent_folder = $parent_id;
+        $sentFolder->save();
+    }
+
+    // Create any additional monitored inbox folders (comma-separated mailbox values)
+    foreach ($inboxFolders as $key => $inboxFolder) {
+        if ($key === 0) {
+            continue;
+        }
+        if ($focus->folderIsRequestTrashOrSent($inboxFolder)) {
+            continue;
+        }
+        $focus->createFolder($inboxFolder, "inbound", $focusUser);
+    }
 } else {
     // Update folders
     require_once("include/SugarFolders/SugarFolders.php");
@@ -423,7 +437,9 @@ if (empty($foldersFoundRow)) {
                 $name = $mod_strings['LNK_MY_DRAFTS'] . ' ('.$stored_options['sentFolder'].')';
                 break;
             case 'sent':
-                $name = $mod_strings['LNK_SENT_EMAIL_LIST'] . ' ('.$stored_options['sentFolder'].')';
+                if (!empty($stored_options['sentFolder'])) {
+                    $name = $mod_strings['LNK_SENT_EMAIL_LIST'] . ' ('.$stored_options['sentFolder'].')';
+                }
                 break;
             case 'archived':
                 $name = $mod_strings['LBL_LIST_TITLE_MY_ARCHIVES'];

--- a/modules/InboundEmail/Save.php
+++ b/modules/InboundEmail/Save.php
@@ -243,7 +243,7 @@ if ($focus->isMailBoxTypeCreateCase() || ($focus->mailbox_type === 'createcase' 
             break;
     }
 } // if
-$storedOptions['folderDelimiter'] = $delimiter;
+$stored_options['folderDelimiter'] = $delimiter;
 
 ////////////////////////////////////////////////////////////////////////////////
 ////    CREATE MAILBOX QUEUE

--- a/modules/InboundEmail/Save.php
+++ b/modules/InboundEmail/Save.php
@@ -330,16 +330,26 @@ if ($type === 'personal' && isset($_REQUEST['account_signature_id']) && $idValid
 
 
 // Folders
-$foldersFound = $focus->db->query('SELECT id FROM folders WHERE folders.id LIKE "'.$focus->id.'"');
+$inboxFolders = array_filter(array_map('trim', explode(',', $focus->mailbox ?: 'INBOX')));
+$inboxFolders = array_values($inboxFolders) ?: ['INBOX'];
+
+$primaryFolder = trim($_REQUEST['primary_folder'] ?? '');
+if (!empty($primaryFolder) && in_array($primaryFolder, $inboxFolders, true) && $inboxFolders[0] !== $primaryFolder) {
+    $inboxFolders = array_values(array_merge(
+        [$primaryFolder],
+        array_filter($inboxFolders, fn($f) => $f !== $primaryFolder)
+    ));
+}
+
+$foldersFound = $focus->db->query('SELECT id, name FROM folders WHERE folders.id LIKE "'.$focus->id.'"');
 $foldersFoundRow = $focus->db->fetchRow($foldersFound);
-$sf = new SugarFolder();
 if (empty($foldersFoundRow)) {
     // Create Folders
     $focusUser = $owner;
     $params = array(
         // Inbox
         "inbound" => array(
-            'name' => $focus->mailbox . ' ('.$focus->name.')',
+            'name' => $inboxFolders[0] . ' ('.$focus->name.')',
             'folder_type' => "inbound",
             'has_child' => 1,
             'dynamic_query' => '',
@@ -369,11 +379,9 @@ if (empty($foldersFoundRow)) {
         ),
     );
 
-
     require_once("include/SugarFolders/SugarFolders.php");
 
     $parent_id = '';
-
 
     foreach ($params as $type => $type_params) {
         if ($type == "inbound") {
@@ -417,7 +425,7 @@ if (empty($foldersFoundRow)) {
         if ($key === 0) {
             continue;
         }
-        if ($focus->folderIsRequestTrashOrSent($inboxFolder)) {
+        if ($focus->folderIsRequestTrashOrSent($inboxFolder, $stored_options['trashFolder'] ?? '', $stored_options['sentFolder'] ?? '')) {
             continue;
         }
         $focus->createFolder($inboxFolder, "inbound", $focusUser);
@@ -425,16 +433,59 @@ if (empty($foldersFoundRow)) {
 } else {
     // Update folders
     require_once("include/SugarFolders/SugarFolders.php");
-    $foldersFound = $focus->db->query('SELECT * FROM folders WHERE folders.id LIKE "'.$focus->id.'" OR '.
-        'folders.parent_folder LIKE "'.$focus->id.'"');
+
+    // Detect and migrate legacy parent folder names that store multiple mailboxes
+    // as a comma-separated string (e.g. "INBOX,Work (AccountName)").
+    // Any mailboxes embedded in the old name that are not already present in
+    // $inboxFolders are merged in so the remainder of the update logic creates
+    // the missing child folder rows.
+    if (!empty($foldersFoundRow['name'])) {
+        $legacyMailboxPart = preg_replace('/\s*\([^)]*\)\s*$/', '', $foldersFoundRow['name']);
+        if (strpos($legacyMailboxPart, ',') !== false) {
+            $systemFolders = [
+                strtolower($stored_options['trashFolder'] ?? ''),
+                strtolower($stored_options['sentFolder'] ?? ''),
+                'sent',
+                'trash',
+            ];
+            $legacyMailboxes = array_filter(array_map('trim', explode(',', $legacyMailboxPart)));
+            foreach ($legacyMailboxes as $legacyMailbox) {
+                if (in_array(strtolower($legacyMailbox), $systemFolders, true)) {
+                    continue;
+                }
+                if (!in_array($legacyMailbox, $inboxFolders, true)) {
+                    $inboxFolders[] = $legacyMailbox;
+                }
+            }
+        }
+    }
+
+    $foldersFound = $focus->db->query(
+        'SELECT * FROM folders WHERE deleted = 0 AND (folders.id LIKE "'.$focus->id.'" OR '.
+        'folders.parent_folder LIKE "'.$focus->id.'")'
+    );
+
+    $inboxNames = array_slice($inboxFolders, 1);
+
     while ($row = $focus->db->fetchRow($foldersFound)) {
         $name = '';
+        $folder = new SugarFolder();
+        $folder->retrieve($row['id']);
+
         switch ($row['folder_type']) {
             case 'inbound':
-                $name = $focus->mailbox . ' ('.$focus->name.')';
+                if (!$row['has_child']) {
+                    if (in_array($row['name'], $inboxNames, true)) {
+                        unset($inboxNames[array_search($row['name'], $inboxNames, true)]);
+                    } else {
+                        $folder->delete();
+                    }
+                    break;
+                }
+                $name = $inboxFolders[0] . ' ('.$focus->name.')';
                 break;
             case 'draft':
-                $name = $mod_strings['LNK_MY_DRAFTS'] . ' ('.$stored_options['sentFolder'].')';
+                $name = $mod_strings['LNK_MY_DRAFTS'];
                 break;
             case 'sent':
                 if (!empty($stored_options['sentFolder'])) {
@@ -446,10 +497,18 @@ if (empty($foldersFoundRow)) {
                 break;
         }
 
-        $folder = new SugarFolder();
-        $folder->retrieve($row['id']);
-        $folder->name = $name;
-        $folder->save();
+        if ($name) {
+            $folder->name = $name;
+            $folder->save();
+        }
+    }
+
+    // Create any newly added monitored inbox folders
+    foreach ($inboxNames as $newInboxFolder) {
+        if ($focus->folderIsRequestTrashOrSent($newInboxFolder, $stored_options['trashFolder'] ?? '', $stored_options['sentFolder'] ?? '')) {
+            continue;
+        }
+        $focus->createFolder($newInboxFolder, "inbound", $owner);
     }
 }
 

--- a/modules/InboundEmail/Save.php
+++ b/modules/InboundEmail/Save.php
@@ -490,6 +490,8 @@ if (empty($foldersFoundRow)) {
             case 'sent':
                 if (!empty($stored_options['sentFolder'])) {
                     $name = $mod_strings['LNK_SENT_EMAIL_LIST'] . ' ('.$stored_options['sentFolder'].')';
+                } else {
+                    $folder->delete();
                 }
                 break;
             case 'archived':

--- a/modules/InboundEmail/Save.php
+++ b/modules/InboundEmail/Save.php
@@ -330,8 +330,8 @@ if ($type === 'personal' && isset($_REQUEST['account_signature_id']) && $idValid
 
 
 // Folders
-$inboxFolders = array_filter(array_map('trim', explode(',', $focus->mailbox ?: 'INBOX')));
-$inboxFolders = array_values($inboxFolders) ?: ['INBOX'];
+$rawMailbox = html_entity_decode($focus->mailbox ?: 'INBOX', ENT_QUOTES | ENT_HTML5, 'UTF-8');
+$inboxFolders = array_values(array_filter(array_map('trim', explode(',', $rawMailbox)))) ?: ['INBOX'];
 
 $primaryFolder = trim($_REQUEST['primary_folder'] ?? '');
 if (!empty($primaryFolder) && in_array($primaryFolder, $inboxFolders, true) && $inboxFolders[0] !== $primaryFolder) {
@@ -359,7 +359,7 @@ if (empty($foldersFoundRow)) {
         ),
         // My Drafts
         "draft" => array(
-            'name' => $mod_strings['LNK_MY_DRAFTS'] . ' ('.$stored_options['sentFolder'].')',
+            'name' => $mod_strings['LNK_MY_DRAFTS'] . ' (' . ($stored_options['sentFolder'] ?? '') . ')',
             'folder_type' => "draft",
             'has_child' => 0,
             'dynamic_query' => '',
@@ -442,15 +442,11 @@ if (empty($foldersFoundRow)) {
     if (!empty($foldersFoundRow['name'])) {
         $legacyMailboxPart = preg_replace('/\s*\([^)]*\)\s*$/', '', $foldersFoundRow['name']);
         if (strpos($legacyMailboxPart, ',') !== false) {
-            $systemFolders = [
-                strtolower($stored_options['trashFolder'] ?? ''),
-                strtolower($stored_options['sentFolder'] ?? ''),
-                'sent',
-                'trash',
-            ];
+            $legacyTrash = $stored_options['trashFolder'] ?? '';
+            $legacySent  = $stored_options['sentFolder'] ?? '';
             $legacyMailboxes = array_filter(array_map('trim', explode(',', $legacyMailboxPart)));
             foreach ($legacyMailboxes as $legacyMailbox) {
-                if (in_array(strtolower($legacyMailbox), $systemFolders, true)) {
+                if ($focus->folderIsRequestTrashOrSent($legacyMailbox, $legacyTrash, $legacySent)) {
                     continue;
                 }
                 if (!in_array($legacyMailbox, $inboxFolders, true)) {
@@ -466,6 +462,7 @@ if (empty($foldersFoundRow)) {
     );
 
     $inboxNames = array_slice($inboxFolders, 1);
+    $sentFolderExists = false;
 
     while ($row = $focus->db->fetchRow($foldersFound)) {
         $name = '';
@@ -485,9 +482,10 @@ if (empty($foldersFoundRow)) {
                 $name = $inboxFolders[0] . ' ('.$focus->name.')';
                 break;
             case 'draft':
-                $name = $mod_strings['LNK_MY_DRAFTS'];
+                $name = $mod_strings['LNK_MY_DRAFTS'] . ' (' . ($stored_options['sentFolder'] ?? '') . ')';
                 break;
             case 'sent':
+                $sentFolderExists = true;
                 if (!empty($stored_options['sentFolder'])) {
                     $name = $mod_strings['LNK_SENT_EMAIL_LIST'] . ' ('.$stored_options['sentFolder'].')';
                 } else {
@@ -503,6 +501,19 @@ if (empty($foldersFoundRow)) {
             $folder->name = $name;
             $folder->save();
         }
+    }
+
+    if (!$sentFolderExists && !empty($stored_options['sentFolder'])) {
+        $sentFolder = new SugarFolder();
+        $sentFolder->name = $mod_strings['LNK_SENT_EMAIL_LIST'] . ' ('.$stored_options['sentFolder'].')';
+        $sentFolder->folder_type = 'sent';
+        $sentFolder->has_child = 0;
+        $sentFolder->dynamic_query = '';
+        $sentFolder->is_dynamic = 1;
+        $sentFolder->created_by = $owner->id;
+        $sentFolder->modified_by = $owner->id;
+        $sentFolder->parent_folder = $focus->id;
+        $sentFolder->save();
     }
 
     // Create any newly added monitored inbox folders

--- a/modules/InboundEmail/ShowInboundFoldersList.php
+++ b/modules/InboundEmail/ShowInboundFoldersList.php
@@ -276,6 +276,7 @@ echo '
         } else {
             var mb = document.getElementById("mailbox");
             mb.value = selectedmbox;
+            mb.dispatchEvent(new Event("change"));
         }
         SUGAR.inboundEmail.listDlg.hide();
      }

--- a/modules/InboundEmail/js/primary_folder.js
+++ b/modules/InboundEmail/js/primary_folder.js
@@ -1,0 +1,101 @@
+/*
+ * SuiteCRM is a customer relationship management program developed by SuiteCRM Ltd.
+ * Copyright (C) 2026 SuiteCRM Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Affero General Public License version 3 as published by the
+ * Free Software Foundation with the addition of the following permission added
+ * to Section 15 as permitted in Section 7(a): FOR ANY PART OF THE COVERED WORK
+ * IN WHICH THE COPYRIGHT IS OWNED BY SUITECRM, SUITECRM DISCLAIMS THE
+ * WARRANTY OF NON INFRINGEMENT OF THIRD PARTY RIGHTS.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * In accordance with Section 7(b) of the GNU Affero General Public License
+ * version 3, these Appropriate Legal Notices must retain the display of the
+ * "Supercharged by SuiteCRM" logo. If the display of the logos is not reasonably
+ * feasible for technical reasons, the Appropriate Legal Notices must display
+ * the words "Supercharged by SuiteCRM".
+ */
+
+(function () {
+    'use strict';
+
+    /**
+     * Hides or shows the primary_folder row (both label and value cells) based on
+     * whether any monitored folders are present.
+     *
+     * @param {HTMLElement} selectEl
+     * @param {boolean}     hasFolders
+     */
+    function setPrimaryFolderVisibility(selectEl, hasFolders) {
+        // The select sits inside a <td>; its preceding sibling <td> is the label cell.
+        var valueCell = selectEl.parentNode;
+        var labelCell = valueCell && valueCell.previousElementSibling;
+        var display   = hasFolders ? '' : 'none';
+        if (valueCell) {
+            valueCell.style.display = display;
+        }
+        if (labelCell) {
+            labelCell.style.display = display;
+        }
+    }
+
+    /**
+     * Populates the primary_folder <select> with options derived from the mailbox field.
+     *
+     * @param {boolean} preserveSelection - true  → keep the value currently shown in the select
+     *                                              (called after the mailbox popup closes).
+     *                                      false → pre-select using the data-selected attribute
+     *                                              set server-side on initial page load.
+     */
+    function updatePrimaryFolderSelect(preserveSelection) {
+        var mailboxEl = document.getElementById('mailbox');
+        var selectEl  = document.getElementById('primary_folder');
+
+        if (!mailboxEl || !selectEl) {
+            return;
+        }
+
+        var folders = (mailboxEl.value || '')
+            .split(',')
+            .map(function (s) { return s.trim(); })
+            .filter(Boolean);
+
+        // Fall back to the first folder when no explicit value is saved yet
+        var targetValue = preserveSelection
+            ? selectEl.value
+            : (selectEl.getAttribute('data-selected') || folders[0] || '');
+
+        selectEl.innerHTML = '';
+
+        folders.forEach(function (folder) {
+            var opt = new Option(folder, folder, false, folder === targetValue);
+            selectEl.add(opt);
+        });
+
+        setPrimaryFolderVisibility(selectEl, folders.length > 0);
+    }
+
+    function init() {
+        var mailboxEl = document.getElementById('mailbox');
+        if (mailboxEl) {
+            mailboxEl.addEventListener('change', function () {
+                updatePrimaryFolderSelect(true);
+            });
+        }
+        updatePrimaryFolderSelect(false);
+    }
+
+    if (document.readyState === 'loading') {
+        document.addEventListener('DOMContentLoaded', init);
+    } else {
+        init();
+    }
+}());

--- a/modules/InboundEmail/language/en_us.lang.php
+++ b/modules/InboundEmail/language/en_us.lang.php
@@ -91,6 +91,7 @@ $mod_strings = array(
     'LBL_CREATE_CASE_REPLY_TEMPLATE' => 'New Case Auto-Reply Template',
     'LBL_CREATE_CASE_REPLY_TEMPLATE_HELP' => 'Select an automated response to notify email senders that a case has been created. The email contains the case number in the Subject line which adheres to the Case Macro setting. This response is only sent when the first email is received from the recipient.',
     'LBL_MAILBOX' => 'Monitored Folders',
+    'LBL_PRIMARY_FOLDER' => 'Primary Folder',
     'LBL_TRASH_FOLDER' => 'Trash Folder',
     'LBL_SENT_FOLDER' => 'Sent Folder',
     'LBL_SELECT' => 'Select',

--- a/modules/InboundEmail/metadata/detailviewdefs.php
+++ b/modules/InboundEmail/metadata/detailviewdefs.php
@@ -146,14 +146,15 @@ $viewdefs ['InboundEmail'] = [
                 ],
                 [
                     'is_ssl',
-                    'trashFolder',
+                    'primary_folder',
                 ],
                 [
                     'connection_string',
-                    'sentFolder'
+                    'trashFolder',
                 ],
                 [
                     'email_body_filtering',
+                    'sentFolder'
                 ],
             ],
             'lbl_outbound_configuration' => [

--- a/modules/InboundEmail/metadata/editviewdefs.php
+++ b/modules/InboundEmail/metadata/editviewdefs.php
@@ -113,7 +113,8 @@ $viewdefs ['InboundEmail'] = [
                                modules/InboundEmail/js/auth_type_fields_toggle.js,
                                modules/InboundEmail/js/owner_toggle.js,
                                modules/InboundEmail/js/test_configuration.js,
-                               modules/InboundEmail/js/panel_toggle.js"}
+                               modules/InboundEmail/js/panel_toggle.js,
+                               modules/InboundEmail/js/primary_folder.js"}
                 </script>
             '
         ],

--- a/modules/InboundEmail/metadata/editviewdefs.php
+++ b/modules/InboundEmail/metadata/editviewdefs.php
@@ -148,26 +148,31 @@ $viewdefs ['InboundEmail'] = [
                     [
                         'name' => 'mailbox',
                         'vname' => 'LBL_MAILBOX',
-                        'customCode' => '<div style="display: flex; align-items:baseline;"><input id="mailbox" name="mailbox" tabindex="90" size="30" maxlength="500" type="text" value="{$fields.mailbox.value}"/> <input type="button" id="subscribeFolderButton" class="button" onclick="openMailboxPopup()" value="{$MOD.LBL_SELECT}"/></div>',
+                        'customCode' => '<div style="display: flex; align-items:baseline;"><input id="mailbox" name="mailbox" tabindex="90" size="30" maxlength="500" type="text" value="{$fields.mailbox.value}" style="cursor: default;" readonly/> <input type="button" id="subscribeFolderButton" class="button" onclick="openMailboxPopup()" value="{$MOD.LBL_SELECT}"/></div>',
                     ]
                 ],
                 [
                     'is_ssl',
+                    [
+                        'name' => 'primary_folder',
+                        'vname' => 'LBL_PRIMARY_FOLDER',
+                        'customCode' => '<select id="primary_folder" name="primary_folder" data-selected="{$fields.primary_folder.value}"></select>',
+                    ]
+                ],
+                [
+                    'connection_string',
                     [
                         'name' => 'trashFolder',
                         'customCode' => '<div style="display: flex; align-items:baseline;"><input name="trashFolder" id="trashFolder" tabindex="92" value="{$fields.trashFolder.value}" size=\'30\' maxlength=\'100\' type="text"/> <input type="button" id="trashFolderButton" class="button" onclick="openTrashMailboxPopup()" value="{$MOD.LBL_SELECT}"/></div>',
                     ]
                 ],
                 [
-                    'connection_string',
+                    'email_body_filtering',
                     [
                         'name' => 'sentFolder',
 
                         'customCode' => '<div style="display: flex; align-items:baseline;"><input id="sentFolder" name="sentFolder" tabindex="95" size="30" maxlength="100" type="text" value="{$fields.sentFolder.value}"/> <input type="button" id="sentFolderButton" class="button" onclick="openSentMailboxPopup()" value="{$MOD.LBL_SELECT}"/></div>',
                     ]
-                ],
-                [
-                    'email_body_filtering',
                 ],
             ],
             'lbl_outbound_configuration' => [

--- a/modules/InboundEmail/vardefs.php
+++ b/modules/InboundEmail/vardefs.php
@@ -410,6 +410,20 @@ $dictionary['InboundEmail'] = [
             'unified_search' => false,
             'comment' => ''
         ],
+        'primary_folder' => [
+            'name' => 'primary_folder',
+            'vname' => 'LBL_PRIMARY_FOLDER',
+            'type' => 'varchar',
+            'len' => '255',
+            'required' => false,
+            'reportable' => false,
+            'massupdate' => false,
+            'inline_edit' => false,
+            'importable' => false,
+            'exportable' => false,
+            'unified_search' => false,
+            'comment' => 'Primary monitored folder used as the parent folder for this inbound email account'
+        ],
         'sentFolder' => [
             'name' => 'sentFolder',
             'vname' => 'LBL_SENT_FOLDER',

